lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -50,43 +55,129 @@ define((require, exports, module) ->
       )
       return
 
-    callbackHidePopover = ->
-      popoverHandler.hide($("#formula"))
-      editor.off("changeSelection", callbackHidePopover)
-      editor.session.off("changeScrollTop", callbackHidePopover)
-      editor.session.off("changeScrollLeft", callbackHidePopover)
-      return
+    jqEditorContainer = $(editor.container)
+    jqFormula = -> $("#formula")
+
+    # ch stands for Context Handler
+    contextHandler = new class
+      @removeRegex: /\\end\{equation\}|\\begin\{equation\}|\\label\{[^\}]*\}/g
+
+      @getEquationRange: (cursorRow) ->
+        i = cursorRow
+        while getContext(editor.session, i - 1) == "equation"
+          i -= 1
+        start = i
+        while getContext(editor.session, i + 1) == "equation"
+          i += 1
+        end = i
+        return [start, end]
+
+      @getWholeEquation: (start, end) ->
+        editor.session.getLines(start, end).join(" ").replace(@removeRegex, "")
+
+      @getPopoverPosition: (row) -> {
+          top: "#{editor.renderer.textToScreenCoordinates(row + 2, 1).pageY}px"
+          left: "#{jqEditorContainer.position().left}px"
+        }
+
+      constructor: ->
+        [@curStart, @curEnd] = [null, null]
+        @currentDelayedUpdateId = null
+        @contextPreviewExists = false
 
-    onLoaded = ->
-      try
-        content = katex.renderToString(
-          editor.getSelectedText(),
+      getCurrentFormula: ->
+        katex.renderToString(
+          @constructor.getWholeEquation(@curStart, @curEnd),
           {displayMode: true}
         )
-      catch e
-        content = e
-      finally
-        popoverHandler.show($("#formula"), content)
-        editor.on("changeSelection", callbackHidePopover)
-        editor.session.on("changeScrollTop", callbackHidePopover)
-        editor.session.on("changeScrollLeft", callbackHidePopover)
-        return
 
-    callback = (editor) -> createPopover(editor)
+      initPopover: =>
+        {row: cursorRow} = editor.getCursorPosition()
+        [@curStart, @curEnd] = @constructor.getEquationRange(cursorRow)
+        popoverPosition = @constructor.getPopoverPosition(@curEnd)
+        try
+          content = @getCurrentFormula()
+        catch e
+          content = e
+        finally
+          popoverHandler.show(jqFormula(), content, popoverPosition)
+
+      updatePopover: ->
+        try
+          content = @getCurrentFormula()
+        catch e
+          content = e
+        finally
+          popoverHandler.setContent(jqFormula(), content)
+
+      delayedUpdatePopover: =>
+        if @currentDelayedUpdateId?
```

Comment:
More resource saving way is to return when there is already timeout in flight. To postpone update if there was any keyboard typing since the last timeout scheduling, you can add a boolean flag, and check it when timeout code runs. If flag is true (user has updated the text) then clear the flag and schedule one more command, otherwise update popover. This way you will schedule 1 command per second, not one command per keyboard hit.